src/app/clusters/door-lock-server/door-lock-server.cpp

```diff
@@ -140,30 +143,115 @@ bool emberAfDoorLockClusterLockDoorCallback(chip::app::CommandHandler * commandO
                                             const chip::app::ConcreteCommandPath & commandPath,
                                             const chip::app::Clusters::DoorLock::Commands::LockDoor::DecodableType & commandData)
 {
-    emberAfDoorLockClusterPrintln("Received Lock Door command (not implemented)");
+    emberAfDoorLockClusterPrintln("Received Lock Door command");
+    bool success = false;
 
-    // TODO: Implement door locking by calling emberAfPluginDoorLockOnDoorLockCommand
+    chip::EndpointId endpoint = commandPath.mEndpointId;
 
-    emberAfSendImmediateDefaultResponse(EMBER_ZCL_STATUS_SUCCESS);
-    return true;
+    bool require_pin = false;
+    Attributes::RequirePINforRemoteOperation::Get(endpoint, &require_pin);
+
+    if(require_pin)
+    {
+        if(!commandData.pinCode.HasValue())
+        {
+            success = false; // Just to be explicit. success == false at this point anywyay
+        }
+        else
+        {
+            const char *cmd_pin = reinterpret_cast<const char*>(commandData.pinCode.Value().data());
+            char *mPin = DoorLockServer::Instance().mPin;
+            if(strncmp(mPin, cmd_pin, 4) == 0)
+            {
+                success = emberAfPluginDoorLockOnDoorLockCommand(endpoint, cmd_pin);
+            }
+            else
+            {
+                success = false;
+            }
+        }
+    }
+    else
+    {
+        const char *cmd_pin = commandData.pinCode.HasValue() ?
+            reinterpret_cast<const char*>(commandData.pinCode.Value().data()) :
+            NULL;
```

Comment:
This does not look right.  Not requiring PIN just means that if it's _not_ provided that's OK and the command executes.  If it's provided, it still needs to be checked.  So the logic in this function should probably test `pinCode.HasValue()` at toplevel, not `require_pin`, and only worry about `require_pin` if there is no PIN.

Reply:
Yea good catch. Reworked the logic.